### PR TITLE
[#128] New patterns

### LINKED ISSUE #128: Add `PatternBuilder` combinators
## preset character classes

- `digit`
- `word`
- `space`
- `upper`
- `lower`
- `alpha`
- `alnum`
- `punct`
- `blank`
- `graph`
- `print`
- `xdigit`
- (character class modifier) `non`

## aliases

- `exactly` = `times`

## new helpers

- `sequence` which takes a non-empty list of `Term`s and returns a `TermSequence` (same idea as doing `PB.then` between each)

### PR DESCRIPTION
**Pull request by skeate**
Closes #128

### CODE REVIEW

File: src/PatternBuilder.ts [hunk 1]
```diff
@@ -47,116 +47,126 @@ type TermSequence = { tag: 'termSequence'; terms: ReadonlyArray<Term> }
 
 type Term = Atom | QuantifiedAtom
 
-type QuantifiedAtom = { tag: 'quantifiedAtom'; atom: Atom; greedy: boolean } & (
-  | { kind: 'star' }
-  | { kind: 'plus' }
+type QuantifiedAtom = { tag: 'quantifiedAtom'; atom: Atom } & (
+  | { kind: 'star'; greedy: boolean }
+  | { kind: 'plus'; greedy: boolean }
```
Comment (skeate) [inline src/PatternBuilder.ts:52]: While the others can _technically_ be marked as non-greedy, I've never found a reason to do so in practice, which is why I left them off of their corresponding constructor functions. Moving this codifies that into the type to allow 100% test coverage

File: src/PatternBuilder.ts [hunk 2]
```diff
@@ -47,116 +47,126 @@ type TermSequence = { tag: 'termSequence'; terms: ReadonlyArray<Term> }
...
-    const lower = typeof c1 === 'string' ? c1.charCodeAt(0) : c1
-    const upper = typeof c2 === 'string' ? c2.charCodeAt(0) : c2
-    return { lower, upper } as const
-  }),
+  ranges: convertRanges(ranges),
 })
 
 /** @since 1.0.0 */
-export const subgroup = (subpattern: Pattern): Atom => ({
+export const subgroup: (subpattern: Pattern) => Atom = subpattern => ({
```
Comment (skeate) [inline src/PatternBuilder.ts:108]: adding explicit type annotations throughout

File: src/PatternBuilder.ts [hunk 3]
```diff
@@ -166,29 +176,44 @@ export const between =
...
-  (alt: TermSequence | Term): TermSequence => ({
-    tag: 'termSequence',
-    terms: alt.tag === 'termSequence' ? [...alt.terms, term] : [alt, term],
-  })
+export const then: (
+  term: Term | TermSequence
+) => (alt: TermSequence | Term) => TermSequence = term => alt => ({
+  tag: 'termSequence',
+  terms: [...getTerms(alt), ...getTerms(term)],
+})
```
Comment (skeate) [inline src/PatternBuilder.ts:199]: found a situation where I wanted to do `TermSequence, then TermSequence` and there's really no reason not to :woman_shrugging:

File: src/PatternBuilder.ts [hunk 4]
```diff
@@ -299,3 +321,73 @@ export const arbitraryFromPattern: (pattern: Pattern) => fc.Arbitrary<string> =
...
+  cc => ({
+    tag: 'atom',
+    kind: 'characterClass',
+    exclude: cc.exclude,
+    ranges: cc.ranges.concat(
+      typeof first === 'string' || first instanceof Array
+        ? convertRanges([first, ...addl])
+        : first.ranges
+    ),
+  })
```
Comment (skeate) [inline src/PatternBuilder.ts:345]: augment character classes with other character classes, or new ranges!

File: src/PatternBuilder.ts (hunk 1 above)
Comment (jacob-alford) [reply src/PatternBuilder.ts:52]: Awesome 👏

**Review by jacob-alford — APPROVED** [review]
Looks great!  One small request, it would be nice to have some short descriptions in the JSDocs above all the exports, because I'm not completely sure what everything does, and the descriptions would help disambiguate things.

Otherwise, looks good!  Thanks for boosting the test coverage 🎉

File: src/PatternBuilder.ts [hunk 5]
```diff
@@ -47,116 +47,174 @@ type TermSequence = { tag: 'termSequence'; terms: ReadonlyArray<Term> }
...
+ */
+export const exactly: (count: number) => (atom: Atom) => QuantifiedAtom = times
 
-/** @since 1.0.0 */
-export const between =
-  (min: number, max: number) =>
-  (atom: Atom): QuantifiedAtom => ({
+/**
+ * Repeat an `Atom` at least some number of times. For example, `atLeast(3)(char('a'))`
+ * represents `aaa`, `aaaaaa`, and `aaaaaaaaaaaaaaaaaaaaaaaa` but not `aa`
```
Comment (jacob-alford) [inline src/PatternBuilder.ts:200]: https://youtu.be/yBLdQ1a4-JI

**Review by jacob-alford — APPROVED** [review]
💯 LGTM

File: src/PatternBuilder.ts (hunk 5 above)
Comment (skeate) [reply src/PatternBuilder.ts:200]: not this one though https://www.youtube.com/watch?v=yBLdQ1a4-JI&t=37s

**Pull request merged**